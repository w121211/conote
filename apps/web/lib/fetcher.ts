import { resolve } from 'path'
import { FetchClient } from './fetcher/fetcher'

const fetchClientPropertyName = `__prevent-name-collision__fetcher`

type GlobalThisWithFetchClient = typeof globalThis & {
  [fetchClientPropertyName]: FetchClient
}

function getFetchClient(): FetchClient {
  if (process.env.NODE_ENV === `production`) {
    return new FetchClient()
  } else {
    const newGlobalThis = globalThis as GlobalThisWithFetchClient
    if (!newGlobalThis[fetchClientPropertyName]) {
      newGlobalThis[fetchClientPropertyName] = new FetchClient(resolve(process.cwd(), '.fetcher_local_cache.dump.json'))
    }
    return newGlobalThis[fetchClientPropertyName]
  }
}

const fetcher = getFetchClient()

export default fetcher
