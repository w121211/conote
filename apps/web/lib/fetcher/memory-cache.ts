/**
 * A simple memory cache, mainly use for avoiding repeatly calls on the same urls during development
 */
import { readFileSync, writeFileSync } from 'fs'
import { LinkScrapedResult } from '../interfaces'

export class MemoryCache {
  private filepath: string
  private cache: Record<string, LinkScrapedResult> = {}

  constructor(cachePath: string) {
    this.filepath = cachePath
    try {
      this.cache = this.load(this.filepath)
    } catch (err) {
      console.log(`cache not found, create a new one`)
      //   console.log('No ')
    }
  }

  private load(filepath: string): Record<string, LinkScrapedResult> {
    console.log(`loading cache from ${filepath}`)
    return JSON.parse(readFileSync(filepath, { encoding: 'utf8' }))
  }

  public dump(): void {
    console.log(`Dump cache to ${this.filepath}`)
    writeFileSync(this.filepath, JSON.stringify(this.cache))
  }

  public get(url: string): LinkScrapedResult {
    if (url in this.cache) {
      console.log('hit cache: ' + url)
      return this.cache[url]
    }
    throw new Error('Cache not found:' + url)
  }

  public set(url: string, result: LinkScrapedResult): void {
    this.cache[url] = result
  }
}
