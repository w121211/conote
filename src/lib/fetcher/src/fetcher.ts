import got from 'got'
// import KeyvRedis from '@keyv/redis'
import { FetchResult, SrcType } from './index'

// const keyv = new KeyvRedis('redis://redis:6379')
// keyv.on('error', (err: any) => console.log('Connection Error', err))

export const fetcher: Record<string, (url: string, domain?: string) => Promise<FetchResult>> = {
  youtube: async function (url: string) {
    const domain = 'youtube'
    const u = new URL(url)
    const vid = new URLSearchParams(u.search).get('v')
    // try {
    // const resp = await got.get('https://youtube.com/get_video_info', { searchParams: { video_id: vid } })
    // } catch (error) {
    //     console.log(error.response.body)
    // }
    const resp = await got.get('https://youtube.com/get_video_info', {
      searchParams: { video_id: vid },
      // cache: keyv,
    })
    const p = new URLSearchParams(resp.body).get('player_response')
    if (p) {
      const j = JSON.parse(p)
      return {
        domain,
        // TODO: 可能會有redirect, short-url
        resolvedUrl: url,
        srcType: SrcType.VIDEO,
        srcId: vid ?? '',
        srcTitle: j.microformat.playerMicroformatRenderer.title.simpleText,
        srcPublishDate: new Date(j.microformat.playerMicroformatRenderer.publishDate).toISOString(),
        authorId: j.microformat.playerMicroformatRenderer.externalChannelId,
        authorName: j.microformat.playerMicroformatRenderer.ownerChannelName,
        keywords: j.videoDetails.keywords,
        description: j.videoDetails.shortDescription,
      }
    }
    throw new Error(`youtube fetcher failed: ${url}`)
  },
  default: async function (url, domain) {
    if (domain === undefined) throw new Error()
    return {
      domain,
      resolvedUrl: url,
      srcType: SrcType.OTHER,
    }
  },
}

// function youtube(): void {
//   console.log('hi')
// }

// const fetchers = [youtube]

// function fetch() {
//   const domain = resolveDomain(url)
//   const fetcher = fetchers.find(e => e.matcher(domain))
//   fetcher
// }
