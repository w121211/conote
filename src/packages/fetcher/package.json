{
  "name": "@conote/fetcher",
  "version": "0.1.0",
  "license": "MIT",
  "scripts": {
    "clean": "rm -rf dist && rm -rf reports",
    "build-ts": "tsc --build",
    "build": "yarn run clean && yarn run build-ts",
    "test": "env-cmd -f ../../web/.env.local jest --verbose --watchAll"
  },
  "main": "./dist/src/index.js",
  "module": "./dist/src/index.js",
  "types": "./dist/src/index.d.ts",
  "devDependencies": {
    "@types/chance": "^1.1.1",
    "@types/jest": "^26.0.20",
    "@types/lodash": "^4.14.155",
    "@types/node": "^13.13.21",
    "@typescript-eslint/eslint-plugin": "^4.16.1",
    "@typescript-eslint/parser": "^4.16.1",
    "encoding": "^0.1.13",
    "env-cmd": "^10.1.0",
    "eslint": "^6.8.0",
    "eslint-config-prettier": "^7.2.0",
    "eslint-plugin-prettier": "^3.3.1",
    "jest": "^25.5.4",
    "prettier": "^2.2.1",
    "ts-jest": "^26.5.3",
    "ts-node": "^8.10.2",
    "ts-node-dev": "^1.0.0-pre.63",
    "typescript": "^4.2.2"
  },
  "dependencies": {
    "@keyv/redis": "^2.1.2",
    "googleapis": "^74.2.0",
    "got": "^11.8.2",
    "keyv": "^4.0.3",
    "metascraper": "^5.24.8",
    "metascraper-author": "^5.24.8"
  },
  "engines": {
    "node": ">=10.0.0"
  }
}
