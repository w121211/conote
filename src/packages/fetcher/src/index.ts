export * from './fetch-client'
export * from './domain/index'
export * from './domain/general'

export type SrcType = 'VIDEO' | 'POST' | 'AUTHOR' | 'OTHER'

export type FetchResult = {
  domain: string
  resolvedUrl: string
  srcId?: string
  srcType: SrcType
  srcTitle?: string
  srcPublishDate?: string
  authorId?: string
  authorName?: string
  keywords?: string[]
  description?: string
}
