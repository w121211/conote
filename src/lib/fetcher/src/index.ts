<<<<<<< HEAD
import { fetcher } from './fetcher'
=======
import * as domainFetcher from './domain-fetcher'
import { LocalCache } from './local-cache'
>>>>>>> backend-dev

export const SrcType = {
  VIDEO: 'VIDEO',
  POST: 'POST',
  AUTHOR: 'AUTHOR',
  OTHER: 'OTHER',
}

<<<<<<< HEAD
export interface FetchResult {
=======
export type FetchResult = {
>>>>>>> backend-dev
  domain: string
  resolvedUrl: string
  srcId?: string
  srcType: string
  srcTitle?: string
  srcPublishDate?: string
  authorId?: string
  authorName?: string
  keywords?: string[]
  description?: string
}

<<<<<<< HEAD
export interface ParsedUrl {
=======
export type ParsedUrl = {
>>>>>>> backend-dev
  url: string
  domain: string
}

export function parseUrl(url: string): ParsedUrl {
<<<<<<< HEAD
  /**
   * TODO: 不同的URL(eg: short-url)可能指向同一頁面，需整合
   * - 特別是URL params
   */
=======
  // TODO: 不同的URL(eg: short-url)可能指向同一頁面，需整合 特別是URL params
>>>>>>> backend-dev
  let u = new URL(url)
  return {
    url: url,
    domain: u.hostname,
  }
}

export async function fetch(url: string): Promise<FetchResult> {
  /** 嘗試連接URL取得來源資訊 */
  const parsed = parseUrl(url)

<<<<<<< HEAD
  // let fetched: FetchResult
  // if (parsed.domain === 'www.youtube.com') {
  //   fetched = await fetcher.youtube(parsed.url)
  // } else {
  //   fetched = await fetcher.default(parsed.url, parsed.domain)
  // }
  // return fetched

  return await fetcher.default(parsed.url, parsed.domain)
=======
  let fetched: FetchResult
  if (parsed.domain === 'www.youtube.com') {
    fetched = await domainFetcher.youtube(parsed.url)
  } else {
    fetched = await domainFetcher.general(parsed.url, parsed.domain)
  }
  return fetched
}

export class FetchClient {
  private cache: LocalCache | null = null

  constructor(localCachePath: string | null = null) {
    if (localCachePath) {
      this.cache = new LocalCache(localCachePath)
    }
  }

  public async fetch(url: string): Promise<FetchResult & { fromCache?: true }> {
    if (this.cache) {
      try {
        return { ...this.cache.get(url), fromCache: true }
      } catch {
        const res = await fetch(url)
        this.cache.set(url, res)
        return res
      }
    }
    return await fetch(url)
  }

  public dump(): void {
    if (this.cache) {
      this.cache.dump()
    }
  }
>>>>>>> backend-dev
}
