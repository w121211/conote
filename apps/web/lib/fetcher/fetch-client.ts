import { MemoryCache } from './memory-cache'
import { tryFetch } from './vendors/index'

export type SrcType = 'VIDEO' | 'POST' | 'AUTHOR' | 'OTHER'

export type FetchResult = {
  domain: string
  finalUrl: string
  srcId?: string
  srcType: SrcType

  // metascraper
  authorId?: string
  authorName?: string
  date?: string
  description?: string
  lang?: string
  title?: string

  // ./packages/scraper
  keywords?: string[]
  tickers?: string[]
  error?: string
}

export class FetchClient {
  private cache: MemoryCache | null = null // cache support

  constructor(localCachePath: string | null = null) {
    if (localCachePath) {
      this.cache = new MemoryCache(localCachePath)
    }
  }

  public async fetch(url: string): Promise<FetchResult & { fromCache?: true }> {
    if (this.cache) {
      try {
        return {
          ...this.cache.get(url),
          fromCache: true,
        }
      } catch {
        const res = await tryFetch(url)
        this.cache.set(url, res)
        return res
      }
    }
    return await tryFetch(url)
  }

  public dump(): void {
    if (this.cache) {
      this.cache.dump()
    }
  }
}
