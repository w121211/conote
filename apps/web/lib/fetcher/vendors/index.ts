import { base } from './base'
import { youtube } from './youtube'
import { FetchResult } from '../fetch-client'

export type DomainFetchFunction = (url: string, domain?: string) => Promise<FetchResult>

export class DomainNotFitError extends Error {
  constructor(message?: string) {
    super(message) // 'Error' breaks prototype chain here
    Object.setPrototypeOf(this, DomainNotFitError.prototype) // restore prototype chain
  }
}

// fallback 依序嘗試執行，若失敗則嘗試下一個
const fetchers: DomainFetchFunction[] = [youtube, base]

export async function tryFetch(url: string): Promise<FetchResult> {
  for (const fetcher of fetchers) {
    try {
      // eslint-disable-next-line no-await-in-loop
      return await fetcher(url)
    } catch (err) {
      if (err instanceof DomainNotFitError) {
        continue
      } else {
        console.error(err)
        break
      }
    }
  }
  throw new Error(`Fetch error: ${url}`)
}
