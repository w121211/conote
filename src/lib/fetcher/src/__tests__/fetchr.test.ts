// import _ from 'lodash'
// import { hash, hashSync } from 'bcryptjs'
// import { PrismaClient } from '@prisma/client'
// import { prisma, foo } from '../context'
<<<<<<< HEAD
import { fetch } from '../index'
import { fetcher } from '../fetcher'
// import { _clean, BOT, TESTUSERS } from '../test-helper'

describe('fetch function', () => {
  it('get fetch url', async () => {
    const res = await fetch('https://test.com')
    expect(res).toMatchSnapshot()
  })
})

describe('fetch youtube', () => {
  it('get meta info', async () => {
    const res = await fetcher.youtube('https://www.youtube.com/watch?v=9KUj1174V8w')
    expect(res).toMatchSnapshot()
=======
import { existsSync, readFileSync, unlinkSync } from 'fs'
import { resolve } from 'path'
import { fetch, FetchClient } from '../index'
import * as domainFetchers from '../domain-fetcher'
// import { _clean, BOT, TESTUSERS } from '../test-helper'

// describe('function fetch', () => {
//   it('resolves url', async () => {
//     expect(await fetch('https://test.com')).toMatchSnapshot()
//   })

//   it('use domain fetch', async () => {
//     expect(await fetch('https://test.com')).toEqual(await domainFetchers.general('https://test.com'))
//     expect(await fetch('https://test.com')).toEqual(await domainFetchers.general('https://test.com'))
//   })
// })

describe('domain fetch', () => {
  it('youtube', async () => {
    expect(await domainFetchers.youtube('https://www.youtube.com/watch?v=9KUj1174V8w')).toMatchSnapshot()
  })

  it('general', async () => {
    expect(await domainFetchers.general('https://test.com', 'test.com')).toMatchSnapshot()
  })
})

describe('class FetchClient', () => {
  const localCachePath = resolve(process.cwd(), './localcache.test.json')
  let fetcher: FetchClient

  // 這個無法使用，可能是因為fetcher是class
  // jest.spyOn(index, 'fetch').mockImplementation(async () => {
  //   console.log('fetch~~~~~')
  //   return {
  //     domain: 'mock.test.com',
  //     resolvedUrl: 'https://mock.test.com',
  //     srcType: 'OTHER',
  //   }
  // })
  // expect(fetchSpy).toBeCalled()

  beforeAll(async () => {
    fetcher = new FetchClient(localCachePath)
    await fetcher.fetch('https://test.com')
  })

  afterAll(() => {
    unlinkSync(localCachePath)
  })

  it('use cache', async () => {
    expect(await fetcher.fetch('https://test.com')).toEqual({
      fromCache: true,
      domain: 'test.com',
      resolvedUrl: 'https://test.com',
      srcType: 'OTHER',
    })
  })

  it('dump cache', async () => {
    fetcher.dump()
    expect(JSON.parse(readFileSync(localCachePath, { encoding: 'utf8' }))).toEqual({
      'https://test.com': { domain: 'test.com', resolvedUrl: 'https://test.com', srcType: 'OTHER' },
    })
>>>>>>> backend-dev
  })
})
