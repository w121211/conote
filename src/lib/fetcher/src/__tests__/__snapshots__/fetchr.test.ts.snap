// Jest Snapshot v1, https://goo.gl/fbAQLP

<<<<<<< HEAD
exports[`Fetch youtube get meta info 1`] = `
Object {
  "authorId": "UCeTEWFsNC3eeUsn9hvsDALQ",
  "authorName": "老汤美股财经",
  "description": "会员加入链接 (福利VIP群)：https://www.youtube.com/channel/UCeTEWFsNC3eeUsn9hvsDALQ/join 
会员福利：
1）日常更新我的交易和【微牛1万投资挑战】
2）每日会员社区总结
3）每月一期Q&A 会员VIP视频
4）VIP 电报群（回答问题）
5）股票分析（股票提示）

老李，小楠，番茄和老汤的「终身学习读书会」终于和大家见面啦！
我们4人会带大家一起读书，一起学习哦～
「终身学习读书会」= 一年100本经典好书+15-20分钟读书音频+干货总结思维导图！
通过微信加入：https://pay.weread.com.tw/redirect/bypass/zhongshenxuexi-wx/
通过网页版加入：https://pay.weread.com.tw/redirect/bypass/zhongshenxuexidushuhui/
通过FB Messenger加入：https://pay.weread.com.tw/redirect/bypass/zhongshenxuexi-fb/
** 加入读书会的用户还可以加入到我们的专属读书会Facebook群组 **
如果遇到任何读书会的问题，可以发邮件到service@vip.weread.tw，客服会尽快为大家解决。

Telegram电报群：https://t.me/joinchat/rJoh6OlfXSIyOWUx
QQ群：519929958
工作邮箱：iamlaotang@gmail.com

微牛证券Webull（美国版需要SSN）：https://act.webull.com/kol-us/share.html?hl=eninviteCode=v7cukN9CV8Zj
用我的链接开户后可获得一支价值在$2.5到$250之间的股票。入金$100后，还可以获得第二支价值$12到$1400的股票哦～

（老虎可以打新哦！）老虎TradeUP（美国版需要SSN）：https://tradeup.marsco.com/accounts?invite=LAOTANG
开户入金任意金额就可获得抽股票机会2次，价值在$2.50 - $1,000之间。入金最高返现200刀哦～

*免责声明*
我（老汤）的全部视频都只是我的个人观点，不是投资建议，我的观点可能是错的，大家在做决定前一定要独立思考。我无法为大家的盈利或损失负责。投资有风险，投资需谨慎！

*Disclaimer*
All my videos are just my own opinions, my opinions could be wrong, my videos are not investment advice, so please make your own decision on any investment, trade, or choice. I am not responsible for your gains or losses. Stock market is risky, so please think twice before you invest.",
  "domain": "youtube",
  "keywords": Array [
    "股票",
    "美股",
    "股市",
    "板块轮动",
    "传统股",
    "价值股",
    "国债收益率",
  ],
  "resolvedUrl": "https://www.youtube.com/watch?v=9KUj1174V8w",
  "srcId": "9KUj1174V8w",
  "srcPublishDate": "2021-03-21T00:00:00.000Z",
  "srcTitle": "【新手系列09】板块轮动买什么股票? 10年期美债收益率为何上涨? 换股前必看!【美股分析】(字幕请点CC)",
  "srcType": "VIDEO",
}
`;

exports[`fetch function get fetch url 1`] = `
=======
exports[`domain fetch general 1`] = `
>>>>>>> backend-dev
Object {
  "domain": "test.com",
  "resolvedUrl": "https://test.com",
  "srcType": "OTHER",
}
`;

<<<<<<< HEAD
exports[`fetch youtube get meta info 1`] = `
=======
exports[`domain fetch youtube 1`] = `
>>>>>>> backend-dev
Object {
  "authorId": "UCeTEWFsNC3eeUsn9hvsDALQ",
  "authorName": "老汤美股财经",
  "description": "会员加入链接 (福利VIP群)：https://www.youtube.com/channel/UCeTEWFsNC3eeUsn9hvsDALQ/join 
会员福利：
1）日常更新我的交易和【微牛1万投资挑战】
2）每日会员社区总结
3）每月一期Q&A 会员VIP视频
4）VIP 电报群（回答问题）
5）股票分析（股票提示）

老李，小楠，番茄和老汤的「终身学习读书会」终于和大家见面啦！
我们4人会带大家一起读书，一起学习哦～
「终身学习读书会」= 一年100本经典好书+15-20分钟读书音频+干货总结思维导图！
通过微信加入：https://pay.weread.com.tw/redirect/bypass/zhongshenxuexi-wx/
通过网页版加入：https://pay.weread.com.tw/redirect/bypass/zhongshenxuexidushuhui/
通过FB Messenger加入：https://pay.weread.com.tw/redirect/bypass/zhongshenxuexi-fb/
** 加入读书会的用户还可以加入到我们的专属读书会Facebook群组 **
如果遇到任何读书会的问题，可以发邮件到service@vip.weread.tw，客服会尽快为大家解决。

Telegram电报群：https://t.me/joinchat/rJoh6OlfXSIyOWUx
QQ群：519929958
工作邮箱：iamlaotang@gmail.com

微牛证券Webull（美国版需要SSN）：https://act.webull.com/kol-us/share.html?hl=eninviteCode=v7cukN9CV8Zj
用我的链接开户后可获得一支价值在$2.5到$250之间的股票。入金$100后，还可以获得第二支价值$12到$1400的股票哦～

（老虎可以打新哦！）老虎TradeUP（美国版需要SSN）：https://tradeup.marsco.com/accounts?invite=LAOTANG
开户入金任意金额就可获得抽股票机会2次，价值在$2.50 - $1,000之间。入金最高返现200刀哦～

*免责声明*
我（老汤）的全部视频都只是我的个人观点，不是投资建议，我的观点可能是错的，大家在做决定前一定要独立思考。我无法为大家的盈利或损失负责。投资有风险，投资需谨慎！

*Disclaimer*
All my videos are just my own opinions, my opinions could be wrong, my videos are not investment advice, so please make your own decision on any investment, trade, or choice. I am not responsible for your gains or losses. Stock market is risky, so please think twice before you invest.",
  "domain": "youtube",
  "keywords": Array [
    "股票",
    "美股",
    "股市",
    "板块轮动",
    "传统股",
    "价值股",
    "国债收益率",
  ],
  "resolvedUrl": "https://www.youtube.com/watch?v=9KUj1174V8w",
  "srcId": "9KUj1174V8w",
  "srcPublishDate": "2021-03-21T00:00:00.000Z",
  "srcTitle": "【新手系列09】板块轮动买什么股票? 10年期美债收益率为何上涨? 换股前必看!【美股分析】(字幕请点CC)",
  "srcType": "VIDEO",
}
`;
